Add Seurat v3 support

mojaveazure opened the pull request:
Hello,

This PR serves to add Seurat v3 support to Conos. I've added new accessor methods for the v3 object, as well as an update to `basicSeuratProc` to support either the v2 or v3 workflow. As an example, the following should work with Seurat v3.0.2 and Conos v1.0.3.9000

```R
library(conos)
panel <-readRDS(file =file.path(find.package(package = "conos"), "extdata","panel.rds"))
panel <-lapply(X = panel, FUN = basicSeuratProc, n.pcs = 10, verbose = FALSE)
con <- Conos$new(panel, n.cores = 2)
con$plotPanel(use.local.clusters = TRUE, title.size = 6, embedding.type = "umap")
con$buildGraph(k = 15, k.self = 5, space = "PCA", ncomps = 30, n.odgenes = 2000,matching.method = "mNN", metric = "angular", score.component.variance = TRUE,verbose = FALSE)
con$findCommunities(method = leiden.community, resolution = 1)
con$plotPanel(font.size = 4, embedding.type = "umap")
con$plotGraph(alpha = 0.1)
con$plotGraph(color.by = "sample", mark.groups = FALSE, alpha = 0.1, show.legend = TRUE)
```

Let me know if you have any questions or concerns; I'd be happy to go over anything I did here in more detail.

VPetukhov commented on R/conos.R:
```diff
@@ -743,6 +782,71 @@ stableTreeClusters <- function(refwt,tests,min.threshold=0.8,min.size=10,n.cores
+      so <- Seurat::FindClusters(object = so, n.iter = 500, n.start = 10, verbose = verbose)
+    }
+    if (tsne) {
+      so <- tryCatch(
```
Don't you think that calling UMAP when `tsne=T` is confusing? Moreover, don't think that everyone prefer this algorithm. So, I think, if one want to run UMAP, they can call `RunUMAP` by hands. Or, if you think it's an often case, you can add option `umap=F` to the function.

mojaveazure replied:
Agreed. Changed with a new `umap` parameter, also enabled use of `umap` for Seurat v2.3.1, when `RunUMAP was introduced, and higher.

VPetukhov replied:
Cool! One more comment about UMAP: you set it by default in `plotSamples` (plot.R) instead of t-SNE. Could you please make behavior consistent there, as well?

VPetukhov commented on DESCRIPTION:
```diff
@@ -1,48 +1,52 @@
 Title: Clustering on Netowrk of Samples
-Version: 1.0.2
-Authors@R: c(person("Peter", "Kharchenko", email = "peter_kharchenko@hms.harvard.edu", role = c("aut", "cre")), 
+Version: 1.0.3.9000
```
As you merge it to master, please, set the version to `1.0.3` and add corresponding note to CHANGELOG.md.

mojaveazure replied:
Done. Let me know if you want me to expand more in the CHANGELOG note

VPetukhov replied:
It's fine, thanks!

mojaveazure commented on R/access_wrappers.R:
```diff
@@ -1,40 +1,168 @@
+  f = 'getOverdispersedGenes',
+  signature = signature('Seurat'),
+  definition = function(sample, n.odgenes = NULL) {
+    if (!requireNamespace('Seurat', quietly = TRUE) || packageVersion('Seurat') < package_version(x = '3.0.0')) {
```
Sounds good. Is there a specific file you use for these utility functions?

VPetukhov replied:
Perhaps, integrations.R suits the best.

mojaveazure replied:
Great, thanks!

mojaveazure commented on R/conos.R:
```diff
@@ -743,6 +782,71 @@ stableTreeClusters <- function(refwt,tests,min.threshold=0.8,min.size=10,n.cores
+    if (verbose) {
+      warning("Running Seurat v2 workflow", call. = FALSE, immediate. = TRUE)
+    }
+    rownames(count.matrix) <- make.unique(rownames(count.matrix))
```
Whoops, left over from my testing. Fixed.

mojaveazure commented on R/conclass.R:
```diff
@@ -199,39 +198,36 @@ Conos <- setRefClass(
-                        l2.sigma=1e5, var.scale =TRUE, ncomps=40, n.odgenes=2000, neighborhood.average=FALSE, neighborhood.average.k=10, matching.mask=NULL,
-                        exclude.samples=NULL, common.centering=TRUE, verbose=TRUE, base.groups=NULL, append.global.axes=TRUE, append.decoys=TRUE, decoy.threshold=1,
-                        n.decoys=k*2, score.component.variance=FALSE, balance.edge.weights=FALSE, balancing.factor.per.cell=NULL, same.factor.downweight=1.0) {
+    buildGraph = function(
```
Sorry, didn't realize that was left over. Reverted back to Conos-style.

mojaveazure commented on R/conos.R:
```diff
@@ -743,6 +782,71 @@ stableTreeClusters <- function(refwt,tests,min.threshold=0.8,min.size=10,n.cores
+#'
+#' @export
+#'
+basicSeuratProc <- function(count.matrix, vars.to.regress=NULL, verbose=TRUE, do.par=TRUE, n.pcs=100, cluster=TRUE, tsne=TRUE) {
```
Didn't see that, sorry. Moved.

mojaveazure commented on R/conos.R:
```diff
@@ -743,6 +782,71 @@ stableTreeClusters <- function(refwt,tests,min.threshold=0.8,min.size=10,n.cores
+  if (!requireNamespace("Seurat", quietly = TRUE)) {
+    stop("You need to install 'Seurat' package to be able to use this function")
+  }
+  if (packageVersion('Seurat') < package_version(x = '3.0.0')) {
```
Sorry, should I make two `basicSeuratProc` functions? One for v2 and one for v3?

VPetukhov replied:
I meant one exported, which calls inside one of the two hidden functions.

mojaveazure replied:
Gotcha. Changed and pushed

mojaveazure commented:
Hi Viktor,

Thanks for the feedback! I've addressed most of your comments, but have a few questions. Where would you like me to put the `checkSeuratV3` function? Do you have a place for utility functions like this? And can you elaborate on what you'd like me to change for the comment
> Better to have two separate functions, wrapped to one.

Thanks again!

VPetukhov commented:
Thank you, Paul!
Please, see the answers above.

mojaveazure commented:
Hi Viktor,

I've addressed those last two comments. Let me know if you see anything else that should be fixed.

VPetukhov commented on R/plot.R:
```diff
@@ -97,26 +97,28 @@ plotSamples <- function(samples, groups=NULL, colors=NULL, gene=NULL, embedding.
+    embedding.type <- if (inherits(x = samples[[1]], what = 'seurat')) {
+      'tsne'
+    } else if (inherits(x = samples[[1]], what = 'Seurat')) {
+      'umap'
```
It's the last part. Please, put tsne here.

mojaveazure replied:
Done

VPetukhov commented:
Thank you again for your cooperation! Hope, users of both Seurat and Conos will be happier now :)
